src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetChunks.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * TextEditingTargetChunks.java
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.studio.client.workbench.views.source.editors.text;
+
+import java.util.ArrayList;
+
+import org.rstudio.core.client.regex.Match;
+import org.rstudio.core.client.regex.Pattern;
+import org.rstudio.studio.client.RStudioGinjector;
+import org.rstudio.studio.client.workbench.prefs.model.UIPrefs;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.LineWidget;
+import org.rstudio.studio.client.workbench.views.source.editors.text.events.EditorModeChangedEvent;
+import org.rstudio.studio.client.workbench.views.source.editors.text.events.ScopeTreeReadyEvent;
+import org.rstudio.studio.client.workbench.views.source.editors.text.rmd.ChunkContextUi;
+import org.rstudio.studio.client.workbench.views.source.editors.text.themes.AceThemes;
+
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
+import com.google.inject.Inject;
+
+public class TextEditingTargetChunks
+             implements PinnedLineWidget.Host,
+                        ScopeTreeReadyEvent.Handler,
+                        EditorModeChangedEvent.Handler
+{
+   public TextEditingTargetChunks(TextEditingTarget target)
+   {
+      target_ = target;
+      toolbars_ = new ArrayList<ChunkContextUi>();
+      initialized_ = false;
+      lastRow_ = 0;
+      renderPass_ = 0;
+      
+      target.getDocDisplay().addScopeTreeReadyHandler(this);
+      target.getDocDisplay().addEditorModeChangedHandler(this);
+      
+      RStudioGinjector.INSTANCE.injectMembers(this);
+   }
+
+   // Public methods ----------------------------------------------------------
+   
+   @Override
+   public void onScopeTreeReady(ScopeTreeReadyEvent event)
+   {
+      if (target_.getDocDisplay().getModeId() == "mode/rmarkdown")
```

Comment:
This isn't something that needs to be done now, but it would be great if we could provide these modes as e.g. an enum (to avoid the stringly-typed ID checks). Ie, we could have a separate class that maps Ace mode IDs to an enum, and have `getModeId()` produce that enum.